test_conformance/spirv_new/test_op_spec_constant.cpp

```diff
@@ -0,0 +1,159 @@
+/******************************************************************
+Copyright (c) 2020 The Khronos Group Inc. All Rights Reserved.
+
+This code is protected by copyright laws and contains material proprietary to
+the Khronos Group, Inc. This is UNPUBLISHED PROPRIETARY SOURCE CODE that may not
+be disclosed in whole or in part to third parties, and may not be reproduced,
+republished, distributed, transmitted, displayed, broadcast or otherwise
+exploited in any manner without the express prior written permission of Khronos
+Group. The receipt or possession of this code does not convey any rights to
+reproduce, disclose, or distribute its contents, or to manufacture, use, or sell
+anything that it may describe, in whole or in part other than under the terms of
+the Khronos Adopters Agreement or Khronos Conformance Test Source License
+Agreement as executed between Khronos and the recipient.
+******************************************************************/
+
+#include "testBase.h"
+#include "types.hpp"
+
+
+template <typename T>
+int run_case(cl_device_id deviceID, cl_context context, cl_command_queue queue,
+             const char *name, T init_buffer, T spec_constant_value,
+             T final_value, bool use_spec_constant,
+             bool (*notEqual)(const T &, const T &) = isNotEqual<T>)
+{
+    clProgramWrapper prog;
+    cl_int err = CL_SUCCESS;
+    if (use_spec_constant)
+    {
+        spec_const new_spec_const =
+            spec_const(101, sizeof(T), &spec_constant_value);
+
+        err =
+            get_program_with_il(prog, deviceID, context, name, new_spec_const);
+    }
+    else
+    {
+        err = get_program_with_il(prog, deviceID, context, name);
+    }
+    SPIRV_CHECK_ERROR(err, "Failed to build program");
+
+    clKernelWrapper kernel = clCreateKernel(prog, "spec_const_kernel", &err);
+    SPIRV_CHECK_ERROR(err, "Failed to create kernel");
+    size_t bytes = sizeof(T);
+    clMemWrapper output_buffer =
+        clCreateBuffer(context, CL_MEM_READ_WRITE | CL_MEM_USE_HOST_PTR, bytes,
+                       &init_buffer, &err);
+    SPIRV_CHECK_ERROR(err, "Failed to create output_buffer");
+
+    err = clSetKernelArg(kernel, 0, sizeof(clMemWrapper), &output_buffer);
+    SPIRV_CHECK_ERROR(err, "Failed to set kernel argument output_buffer");
+
+    size_t work_size = 1;
+    err = clEnqueueNDRangeKernel(queue, kernel, 1, NULL, &work_size, NULL, 0,
+                                 NULL, NULL);
+    SPIRV_CHECK_ERROR(err, "Failed to enqueue kernel");
+    clFinish(queue);
```

Comment:
Minor: consider checking the return value from `clFinish`.  Or, could it be omitted entirely?  The subsequent call to `clEnqueueReadBuffer` is blocking.